File: src/lib/server/services/pacenote.service.ts
/**
 * PaceNote Service
 * 
 * Handles generation of professional pace notes for CAF members using Workers AI.
 * Integrates rank-specific competencies and structured feedback templates.
 */

import { createWorkersAIService, type WorkersAIService, type WorkersAIMessage } from './workers-ai.service.js';
import type { WorkersAIResponse } from '../types/index.js';

export interface PaceNoteInput {
	rank: 'Cpl' | 'MCpl' | 'Sgt' | 'WO';
	memberDetails: string;
	observations: string;
	competencyFocus?: string[];
}

export interface PaceNoteOutput {
	feedback: string;
	rank: string;
	generatedAt: Date;
	usage: {
		tokens: number;
		cost: number;
	};
}

/**
 * Base prompt template for pace note generation
 */
const BASE_PROMPT_TEMPLATE = `You are a CAF (Canadian Armed Forces) supervisor writing a professional pace note for a {RANK}.

INSTRUCTIONS:
- Write a professional, constructive feedback note in two paragraphs
- First paragraph: Describe the events and observations objectively
- Second paragraph: Focus on outcomes, impact, and specific competencies demonstrated
- Use appropriate CAF terminology and maintain professional tone
- Be specific and actionable in your feedback
- Focus on the competencies relevant to the {RANK} rank level

RANK-SPECIFIC COMPETENCIES FOR {RANK}:
{COMPETENCIES}

MEMBER DETAILS:
{MEMBER_DETAILS}

OBSERVATIONS:
{OBSERVATIONS}

Generate a professional pace note based on the above information:`;

/**
 * Rank-specific competency frameworks
 */
const RANK_COMPETENCIES = {
	'Cpl': [
		'Technical proficiency in trade skills',
		'Following orders and procedures',
		'Basic leadership in small team settings',
		'Personal responsibility and accountability',
		'Professional development and learning',
		'Safety awareness and compliance'
	],
	'MCpl': [
		'Team leadership and mentoring',
		'Technical expertise and problem-solving',
		'Training and developing junior members',
		'Resource management and planning',
		'Communication and coordination',
		'Initiative and decision-making in routine situations'
	],
	'Sgt': [
		'Supervision and management of personnel',
		'Strategic thinking and planning',
		'Advanced technical knowledge and instruction',
		'Performance management and development',
		'Inter-departmental coordination',
		'Leadership in complex or challenging situations'
	],
	'WO': [
		'Senior leadership and organizational impact',
		'Policy development and implementation',
		'Mentoring and developing leaders',
		'Strategic planning and resource allocation',
		'Representing the organization externally',
		'Leading organizational change and improvement'
	]
};

export class PaceNoteService {
	private aiService: WorkersAIService;

	constructor(ai: Ai) {
		this.aiService = createWorkersAIService(ai, {
			model: '@cf/meta/llama-3.1-8b-instruct',
			maxTokens: 800,
			temperature: 0.7,
			topP: 0.9
		});
	}

	/**
	 * Generate a pace note based on input parameters
	 */
	async generatePaceNote(input: PaceNoteInput): Promise<PaceNoteOutput> {
		try {
			// Validate input
			this.validateInput(input);

			// Build the prompt
			const prompt = this.buildPrompt(input);

			// Generate the pace note using AI
			const response = await this.aiService.generateFromPrompt(prompt);

			return {
				feedback: response.response,
				rank: input.rank,
				generatedAt: new Date(),
				usage: {
					tokens: response.usage?.total_tokens || 0,
					cost: response.cost || 0
				}
			};

		} catch (error) {
			// Re-throw with context
			if (error && typeof error === 'object' && 'code' in error) {
				throw error; // WorkersAI error, pass through
			}
			
			throw {
				code: 'PACENOTE_ERROR',
				message: 'Failed to generate pace note',
				details: { originalError: error }
			};
		}
	}

	/**
	 * Validate pace note input
	 */
	private validateInput(input: PaceNoteInput): void {
		if (!input.rank || !['Cpl', 'MCpl', 'Sgt', 'WO'].includes(input.rank)) {
			throw {
				code: 'INVALID_RANK',
				message: 'Rank must be one of: Cpl, MCpl, Sgt, WO'
			};
		}

		if (!input.memberDetails || input.memberDetails.trim().length < 10) {
			throw {
				code: 'INVALID_MEMBER_DETAILS',
				message: 'Member details must be at least 10 characters long'
			};
		}

		if (!input.observations || input.observations.trim().length < 20) {
			throw {
				code: 'INVALID_OBSERVATIONS',
				message: 'Observations must be at least 20 characters long'
			};
		}

		// Check for reasonable length limits
		if (input.memberDetails.length > 1000) {
			throw {
				code: 'MEMBER_DETAILS_TOO_LONG',
				message: 'Member details must be less than 1000 characters'
			};
		}

		if (input.observations.length > 2000) {
			throw {
				code: 'OBSERVATIONS_TOO_LONG',
				message: 'Observations must be less than 2000 characters'
			};
		}
	}

	/**
	 * Build the complete prompt for pace note generation
	 */
	private buildPrompt(input: PaceNoteInput): string {
		const competencies = RANK_COMPETENCIES[input.rank];
		const competencyList = competencies
			.map((comp, index) => `${index + 1}. ${comp}`)
			.join('\n');

		let prompt = BASE_PROMPT_TEMPLATE
			.replace(/\{RANK\}/g, input.rank)
			.replace('{COMPETENCIES}', competencyList)
			.replace('{MEMBER_DETAILS}', input.memberDetails.trim())
			.replace('{OBSERVATIONS}', input.observations.trim());

		// Add specific competency focus if provided
		if (input.competencyFocus && input.competencyFocus.length > 0) {
			const focusAreas = input.competencyFocus.join(', ');
			prompt += `\n\nSPECIFIC COMPETENCY FOCUS: Pay particular attention to ${focusAreas}`;
		}

		return prompt;
	}

	/**
	 * Get available ranks
	 */
	getAvailableRanks(): Array<{ value: string; label: string; description: string }> {
		return [
			{
				value: 'Cpl',
				label: 'Corporal (Cpl)',
				description: 'Junior NCO - Team leadership and technical proficiency'
			},
			{
				value: 'MCpl',
				label: 'Master Corporal (MCpl)', 
				description: 'Experienced NCO - Team mentoring and training'
			},
			{
				value: 'Sgt',
				label: 'Sergeant (Sgt)',
				description: 'Senior NCO - Supervision and management'
			},
			{
				value: 'WO',
				label: 'Warrant Officer (WO)',
				description: 'Senior leader - Strategic planning and organizational impact'
			}
		];
	}

	/**
	 * Get competencies for a specific rank
	 */
	getCompetenciesForRank(rank: keyof typeof RANK_COMPETENCIES): string[] {
		return [...RANK_COMPETENCIES[rank]];
	}

	/**
	 * Update AI service configuration
	 */
	updateAIConfig(config: Parameters<WorkersAIService['updateConfig']>[0]): void {
		this.aiService.updateConfig(config);
	}

	/**
	 * Get current AI configuration
	 */
	getAIConfig(): ReturnType<WorkersAIService['getConfig']> {
		return this.aiService.getConfig();
	}
}

/**
 * Factory function to create PaceNoteService instance
 */
export function createPaceNoteService(ai: Ai): PaceNoteService {
	return new PaceNoteService(ai);
}

File: internal/services/pacenote_service.go
package services

import (
	"context"
	"fmt"
	"strings"
	"time"

	"caf-gpt/internal/storage"
)

// PaceNoteRank represents valid CAF ranks
type PaceNoteRank string

const (
	RankCpl  PaceNoteRank = "Cpl"
	RankMCpl PaceNoteRank = "MCpl"
	RankSgt  PaceNoteRank = "Sgt"
	RankWO   PaceNoteRank = "WO"
)

// RankInfo contains rank display information
type RankInfo struct {
	Value string `json:"value"`
	Label string `json:"label"`
}

// PaceNoteInput represents input for pace note generation
type PaceNoteInput struct {
	Rank            PaceNoteRank `json:"rank"`
	Observations    string       `json:"observations"`
	CompetencyFocus []string     `json:"competency_focus,omitempty"`
}

// PaceNoteOutput represents the generated pace note
type PaceNoteOutput struct {
	Feedback    string    `json:"feedback"`
	Rank        string    `json:"rank"`
	GeneratedAt time.Time `json:"generated_at"`
	Usage       Usage     `json:"usage"`
}

// PaceNoteService handles pace note generation
type PaceNoteService struct {
	openRouter *OpenRouterClient
	storage    *storage.TigrisClient
	model      string
}

// NewPaceNoteService creates a new pace note service
func NewPaceNoteService(openRouter *OpenRouterClient, storage *storage.TigrisClient, model string) *PaceNoteService {
	return &PaceNoteService{
		openRouter: openRouter,
		storage:    storage,
		model:      model,
	}
}

// GetAvailableRanks returns the list of available ranks
func (s *PaceNoteService) GetAvailableRanks() []RankInfo {
	return []RankInfo{
		{Value: "Cpl", Label: "Corporal (Cpl)"},
		{Value: "MCpl", Label: "Master Corporal (MCpl)"},
		{Value: "Sgt", Label: "Sergeant (Sgt)"},
		{Value: "WO", Label: "Warrant Officer (WO)"},
	}
}

// ValidateInput validates pace note input
func (s *PaceNoteService) ValidateInput(input *PaceNoteInput) error {
	validRanks := map[PaceNoteRank]bool{
		RankCpl: true, RankMCpl: true, RankSgt: true, RankWO: true,
	}

	if !validRanks[input.Rank] {
		return fmt.Errorf("invalid rank: %s", input.Rank)
	}

	if len(strings.TrimSpace(input.Observations)) < 20 {
		return fmt.Errorf("observations must be at least 20 characters long")
	}

	if len(input.Observations) > 2000 {
		return fmt.Errorf("observations must be less than 2000 characters")
	}

	return nil
}

// GeneratePaceNote generates a pace note based on input
func (s *PaceNoteService) GeneratePaceNote(ctx context.Context, input *PaceNoteInput) (*PaceNoteOutput, error) {
	// Validate input
	if err := s.ValidateInput(input); err != nil {
		return nil, fmt.Errorf("validation error: %w", err)
	}

	// Build system prompt
	systemPrompt, err := s.buildSystemPrompt(ctx, input.Rank, input.CompetencyFocus)
	if err != nil {
		return nil, fmt.Errorf("failed to build system prompt: %w", err)
	}

	// Generate using OpenRouter
	response, err := s.openRouter.GenerateFromPrompt(
		ctx,
		s.model,
		systemPrompt,
		strings.TrimSpace(input.Observations),
		0.1, // Low temperature for consistent output
	)
	if err != nil {
		return nil, fmt.Errorf("failed to generate pace note: %w", err)
	}

	if len(response.Choices) == 0 {
		return nil, fmt.Errorf("no response generated")
	}

	return &PaceNoteOutput{
		Feedback:    response.Choices[0].Message.Content,
		Rank:        string(input.Rank),
		GeneratedAt: time.Now(),
		Usage: Usage{
			PromptTokens:     response.Usage.PromptTokens,
			CompletionTokens: response.Usage.CompletionTokens,
			TotalTokens:      response.Usage.TotalTokens,
		},
	}, nil
}

// buildSystemPrompt builds the system prompt with competencies and examples
func (s *PaceNoteService) buildSystemPrompt(ctx context.Context, rank PaceNoteRank, competencyFocus []string) (string, error) {
	// Read base prompt template
	basePrompt, err := s.readPromptFile(ctx, "pacenote/prompts/base.md")
	if err != nil {
		return "", fmt.Errorf("failed to read base prompt: %w", err)
	}

	// Get competencies for rank
	competencies, err := s.getCompetenciesForRank(ctx, rank)
	if err != nil {
		return "", fmt.Errorf("failed to get competencies: %w", err)
	}

	// Get examples
	examples, err := s.getExamples(ctx)
	if err != nil {
		return "", fmt.Errorf("failed to get examples: %w", err)
	}

	// Replace placeholders in base template
	systemPrompt := strings.ReplaceAll(basePrompt, "{{competency_list}}", competencies)
	systemPrompt = strings.ReplaceAll(systemPrompt, "{{examples}}", examples)

	// Add specific competency focus if provided
	if len(competencyFocus) > 0 {
		focusAreas := strings.Join(competencyFocus, ", ")
		systemPrompt += fmt.Sprintf("\n\nSPECIFIC COMPETENCY FOCUS: Pay particular attention to %s", focusAreas)
	}

	return systemPrompt, nil
}

// readPromptFile reads a prompt file from embedded prompts
func (s *PaceNoteService) readPromptFile(ctx context.Context, path string) (string, error) {
	// For now, we'll read from storage. In the final version, this will be embedded
	return s.storage.ReadFile(ctx, path)
}

// getCompetenciesForRank gets competencies for a specific rank
func (s *PaceNoteService) getCompetenciesForRank(ctx context.Context, rank PaceNoteRank) (string, error) {
	filePath := fmt.Sprintf("paceNote/%s.md", strings.ToLower(string(rank)))
	return s.storage.ReadFile(ctx, filePath)
}

// getExamples gets example pace notes
func (s *PaceNoteService) getExamples(ctx context.Context) (string, error) {
	return s.storage.ReadFile(ctx, "paceNote/examples.md")
}